### PR TITLE
Graceful shutdown strategy used as default one

### PR DESCRIPTION
**Pull request by JiriOndrusek**
fixes https://github.com/apache/camel-quarkus/issues/3179

Adds following behavior:
- in development mode, if no graceful shutdown timeout is set, there is no change (`noShutdownStrategy` is used)
- in other cases `gracefulShutdownStrategy` (same a is used in the Camel) is used
@ppalaga I think that this should be mentioned in some doc, do you have a suggestion where to write it?

### CODE REVIEW

**Review by oscerd — APPROVED** [review]

File: extensions-core/core/deployment/src/main/java/org/apache/camel/quarkus/core/deployment/CamelContextProcessor.java [hunk 1]
```diff
@@ -102,6 +106,28 @@ CamelContextBuildItem context(
...
+            CamelContextRecorder recorder,
+            Capabilities capabilities,
+            BuildProducer<CamelContextCustomizerBuildItem> producer) {
+        if (capabilities.isPresent(CamelCapabilities.MAIN)) {
+            String val = CamelSupport.getOptionalConfigValue("camel.main.shutdownTimeout", String.class, null);
+            if (val == null) {
+                //if no graceful timeout is set in development mode, graceful shutdown is replaced with no shutdown
+                producer.produce(new CamelContextCustomizerBuildItem(recorder.createNoShutdownStrategyCustomizer()));
+            }
+        }
```
Comment (aldettinger) [inline extensions-core/core/deployment/src/main/java/org/apache/camel/quarkus/core/deployment/CamelContextProcessor.java:128]: Shouldn't we also set no shutdown strategy in DEV mode when no running with main ?

**Comment by aldettinger** [comment]
@JiriOndrusek Looks good, still few questions. Logically, it should make `quarkus.camel.main.shutdown.timeout=xxx` work. I don't see how to automatic test this but wonder if you have been able to check this manually on your machine ?

**Comment by jamesnetherton** [comment]
> I don't see how to automatic test this

As a minimum, we could probably use [`QuarkusDevModeTest`](https://github.com/apache/camel-quarkus/blob/5ad13f2be3d6d8ccb771dcea4c37a7d0c600b24e/integration-tests/main-devmode/src/test/java/org/apache/camel/quarkus/main/CamelMainRoutesIncludePatternWithNoPrefixDevModeTest.java#L41) and verify that the correct shutdown strategy is configured on the `CamelContext`, depending on whether `camel.main.shutdownTimeout` is set.

File: extensions-core/core/deployment/src/main/java/org/apache/camel/quarkus/core/deployment/CamelContextProcessor.java (hunk 1 above)
Comment (JiriOndrusek) [reply extensions-core/core/deployment/src/main/java/org/apache/camel/quarkus/core/deployment/CamelContextProcessor.java:128]: makes sense, I'll fix it.

File: extensions-core/core/deployment/src/main/java/org/apache/camel/quarkus/core/deployment/CamelContextProcessor.java [hunk 2]
```diff
@@ -102,6 +106,28 @@ CamelContextBuildItem context(
...
+    /**
+     * This step customizes camel context for development mode.
+     *
+     * @param recorder     the recorder
+     * @param capabilities the registered capabilities
+     * @param producer     producer of context customizer build item
+     */
+    @Record(ExecutionTime.STATIC_INIT)
+    @BuildStep(onlyIf = IsDevelopment.class)
+    public void developmenMode(
```
[the remark this answers is not in the document]
Comment (JiriOndrusek) [reply extensions-core/core/deployment/src/main/java/org/apache/camel/quarkus/core/deployment/CamelContextProcessor.java:118]: I'll fix it.

**Comment by JiriOndrusek** [comment]
I'll try to add a test.
(@aldettinger I tested it locally with an app)
What do you think about a doc change regarding this?

File: extensions-core/core/deployment/src/main/java/org/apache/camel/quarkus/core/deployment/CamelContextProcessor.java (hunk 1 above)
Comment (JiriOndrusek) [reply extensions-core/core/deployment/src/main/java/org/apache/camel/quarkus/core/deployment/CamelContextProcessor.java:124]: fixed

File: extensions-core/core/deployment/src/main/java/org/apache/camel/quarkus/core/deployment/CamelContextProcessor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (JiriOndrusek) [reply extensions-core/core/deployment/src/main/java/org/apache/camel/quarkus/core/deployment/CamelContextProcessor.java]: fixed

**Comment by JiriOndrusek** [comment]
@jamesnetherton , @aldettinger I added 3 tests:

1. in core (not dev mode) - verifying that `DefaultShutdownStrategy` is used.
2. in main-devmode - `CamelContextNoShutdownStrategyDevModeTest` -  if `camel.main.shutdownTimeout` is set, `DefaulShutdownStrategy` is used 
3. in main-devmode - `CamelContextNoShutdownStrategyDevModeTest` -  if `camel.main.shutdownTimeout` is not set, `NoShutdownStrategy` is used

**Comment by aldettinger** [comment]
@JiriOndrusek It looks better :+1: 

Do we have a test for core in DEV mode ? I mean without using camel-main in DEV mode. My current reading is that we will use the default shutdown strategy whereas no shutdown strategy would be preferrable ?

About the doc change: Are you talking about stating in the doc that the shutdown strategy has changed ? IMHO, I don't think it is needed.

**Comment by JiriOndrusek** [comment]
@aldettinger I would say that the use of `DefaultShudownStrategy` is preferable in all cases, with one exception - dev mode - because of quick redeploy. (but if in dev-mode `camel.main.shutdownTimeout` is set, default shutdown strategy will be used)

As for documentation. Current status was, that `NoShutdownStrategy` was used in all cases. This fix brings a change, which could affect some users. This is the reason, why I would say that doc should be updated - somewhere. (that default shutdown strategy was changed - mainly for not dev-mode)

**Comment by aldettinger** [comment]
> I would say that the use of DefaultShudownStrategy is preferable in all cases, with one exception - dev mode - because of quick redeploy. (but if in dev-mode camel.main.shutdownTimeout is set, default shutdown strategy will be used)

I agree. Maybe I missread that with the current implementation, `DefaultShudownStrategy` is used in DEV mode when camel main is disabled. Do you agree with that ?

> As for documentation. Current status was, that NoShutdownStrategy was used in all cases. This fix brings a change, which could affect some users. This is the reason, why I would say that doc should be updated - somewhere. (that default shutdown strategy was changed - mainly for not dev-mode)

Ok, it's clearer now. I think James has an idea that this could be in the migration guide. I think it is a good option.

**Comment by JiriOndrusek** [comment]
@aldettinger 
```
DefaultShudownStrategy is used in DEV mode when camel main is disabled. Do you agree with that ?
```
Not entirely correct. `DefaultShutdownStrategy` is used in dev mode only if `camel.main.shutdownTimeout` is not null. (without taking into account whether camel main is disabled) - may not be the solution which we want. So in case that different condition should be applied, I'll change it.

**Comment by JiriOndrusek** [comment]
I hope that following statement will make the current behavior clear.

Current implementation is using only `NoShutdownStrategy`.
 This fix changes strategy to `DefaultShutdownStrategy`,
Only if `camel.main.shutdownTimeout` is not set and dev mode is enabled -  `NoShutdownStrategy` remains.

**Comment by JiriOndrusek** [comment]
I'll add change into the migration doc -  https://github.com/apache/camel-quarkus/tree/main/docs/modules/ROOT/pages/migration-guide

**Review by aldettinger — APPROVED** [review]
@JiriOndrusek I'm +1 on adding the documentation to the migration guide

And indeed, I see it that now it does not take into account whether camel main is disabled. So the test coverage is good :+1:

**Comment by JiriOndrusek** [comment]
@aldettinger documentation is added.

**Comment by aldettinger** [comment]
The migration guide is definitely the right place. We could merge this PR from my point of view :+1:

**Review by ppalaga — APPROVED** [review]

**Review by jamesnetherton — APPROVED** [review]

**Pull request merged**